src/librustc_typeck/astconv.rs

```diff
@@ -1225,10 +1225,15 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
 
     /// Transform a `PolyTraitRef` into a `PolyExistentialTraitRef` by
     /// removing the dummy `Self` type (`trait_object_dummy_self`).
-    fn trait_ref_to_existential(&self, trait_ref: ty::TraitRef<'tcx>)
-                                -> ty::ExistentialTraitRef<'tcx> {
+    fn trait_ref_to_existential(
+        &self,
+        trait_ref: ty::TraitRef<'tcx>,
+    ) -> ty::ExistentialTraitRef<'tcx> {
         if trait_ref.self_ty() != self.tcx().types.trait_object_dummy_self {
-            bug!("trait_ref_to_existential called on {:?} with non-dummy Self", trait_ref);
+            self.tcx().sess.delay_span_bug(DUMMY_SP, &format!(
+                "trait_ref_to_existential called on {:?} with non-dummy Self",
+                trait_ref,
+            ));
```

Comment:
I don't think this is the correct fix - the caller shouldn't pass in a bad `TraitRef` in the first place.

Reply:
@eddyb Isn't this a reasonable as a way to paper over the issue while the underlying cause is being researched?

Reply:
Oh I thought this code was in `librustc`.

Looking closer at the code, I don't think `trait_ref_to_existential` should exist as a method.
It should be a closure defined just before its two callsites, and it could refer to the `dummy_self` variable in scope directly.

Also, this definitely needs a FIXME. The overall bug appears to be a missing filter on top of `expand_trait_aliases`, which picks up non-`supertraits` where clauses - but also, the object safety completely ignores trait aliases, which could be object safety hazards.

So trait aliases aren't fully implemented yet, I'm guessing?

Reply:
cc @alexreg 

If you want to paper over something, please file a new issue for that FIXME and label it with `F-trait_alias` so we won't miss it.

Reply:
Created https://github.com/rust-lang/rust/issues/66420 and added `FIXME` to the code.

Reply:
Yeah, this seems like a reasonable temporary fix, but ideally we want something proper in the medium term.

Reply:
@eddyb Can you clarify what you mean about object safety? As far as I can tell, this is already correctly banned (both the existing `trait-alias-object-fail` test and a quick try on the Playground seem to confirm this).